2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/scalastream/ResponseHandler.scala

```diff
@@ -96,122 +97,141 @@ class ResponseHandler(config: CollectorConfig, sinks: CollectorSinks)(implicit c
+      // Split events into Good and Bad
+      val eventSplit = SplitBatch.splitAndSerializePayload(event, sinks.good.MaxBytes)
 
-        // Send events to respective sinks
-        val sinkResponseGood = sinks.good.storeRawEvents(eventSplit.good, partitionKey)
-        val sinkResponseBad  = sinks.bad.storeRawEvents(eventSplit.bad, partitionKey)
+      // Send events to respective sinks
+      val sinkResponseGood = sinks.good.storeRawEvents(eventSplit.good, partitionKey)
+      val sinkResponseBad = sinks.bad.storeRawEvents(eventSplit.bad, partitionKey)
 
-        // Sink Responses for Test Sink
-        sinkResponseGood ++ sinkResponseBad
-      } else {
-        null
-      }
-
-      val policyRef = config.p3pPolicyRef
-      val CP = config.p3pCP
-
-      val headersWithoutCookie = List(
-        RawHeader("P3P", "policyref=\"%s\", CP=\"%s\"".format(policyRef, CP)),
-        getAccessControlAllowOriginHeader(request),
-        `Access-Control-Allow-Credentials`(true)
-      )
-
-      val headers = config.cookieConfig match {
-        case Some(cookieConfig) =>
-          val responseCookie = HttpCookie(
-            cookieConfig.name, networkUserId,
-            expires=Some(DateTime.now + cookieConfig.expiration),
-            domain=cookieConfig.domain
-          )
-          `Set-Cookie`(responseCookie) :: headersWithoutCookie
-        case None => headersWithoutCookie
-      }
+      // Sink Responses for Test Sink
+      sinkResponseGood ++ sinkResponseBad
+    } else {
+      Nil
+    }
 
-      val (httpResponse, badQsResponse) = if (path startsWith "/r/") {
-        // A click redirect
-        try {
-          // TODO: log errors to Kinesis as BadRows
-          val target = URLEncodedUtils.parse(URI.create("?" + queryParams), "UTF-8")
-            .find(_.getName == "u")
-            .map(_.getValue)
-          target match {
-            case Some(t) => HttpResponse(302).withHeaders(`Location`(t) :: headers) -> Nil
-            // case None => badRequest -> sinks.bad.storeRawEvents(List("TODO".getBytes), partitionKey)
-            case None => {
-              val everythingSerialized = new String(SplitBatch.ThriftSerializer.get().serialize(event))
-              badRequest -> sinks.bad.storeRawEvents(List(createBadRow(event, s"Redirect failed due to lack of u parameter")), partitionKey)
-            }
-          }
-        } catch {
-          case NonFatal(e) => {
+    val headers = composeHeaders(request, shouldRedirect, networkUserId, queryParams)
+
+    val (httpResponse, badQsResponse) = if (path startsWith "/r/") {
+      // A click redirect
+      try {
+        // TODO: log errors to Kinesis as BadRows
+        val target = URLEncodedUtils.parse(URI.create("?" + queryParams), "UTF-8")
+          .find(_.getName == "u")
+          .map(_.getValue)
+        target match {
+          case Some(t) => HttpResponse(302).withHeaders(`Location`(t) :: headers) -> Nil
```

Comment:
What happens if we have two location headers because of https://github.com/snowplow/snowplow/pull/2755/files#diff-326d86609fbd1260182e8d32c3e165a0R231 ?